### PR TITLE
fix #1683, Allow metrics to be aggregated before sending to UI

### LINKED ISSUE #1683: Allow metrics to be aggregated before sending to UI

### PR DESCRIPTION
**Pull request by clockfly**

### CODE REVIEW

**Comment by stanleyxu2005** [comment]
Lgtm +1

**Comment by huafengw** [comment]
HistoryMetricsServiceSpec failed

**Comment by clockfly** [comment]
updated

File: core/src/main/scala/io/gearpump/metrics/MetricsAggregator.scala [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package io.gearpump.metrics
```
Comment (manuzhang) [inline core/src/main/scala/io/gearpump/metrics/MetricsAggregator.scala]: license header

File: core/src/main/scala/io/gearpump/util/HistoryMetricsService.scala [hunk 2]
```diff
@@ -80,11 +82,26 @@ class HistoryMetricsService(name: String, config: HistoryMetricsConfig) extends
...
+             aggregators += aggregatorClazz -> aggregatorTry.get
+           } else {
+             LOG.error("Cannot apply aggregator " + aggregatorClazz, aggregatorTry.failed.get)
+             aggregators += aggregatorClazz -> new SkipAllAggregator
+           }
+         }
+         aggregators(aggregatorClazz)
+       }
+
+      sender ! HistoryMetrics(inputPath, aggregator.aggregate(fetchMetricsHistory(inputPath, readLatest)))
```
Comment (manuzhang) [inline core/src/main/scala/io/gearpump/util/HistoryMetricsService.scala]: what if aggregate fails ?

Comment (clockfly) [inline core/src/main/scala/io/gearpump/util/HistoryMetricsService.scala]: If exception happen, let it throw.

Comment (manuzhang) [inline core/src/main/scala/io/gearpump/util/HistoryMetricsService.scala]: Then why should Aggregator instantiation failure should be treated differently ?

Comment (clockfly) [inline core/src/main/scala/io/gearpump/util/HistoryMetricsService.scala]: one is expected, the other is not expected.

Comment (clockfly) [inline core/src/main/scala/io/gearpump/util/HistoryMetricsService.scala]: Typically, I will just say "fail fast", most of the codebase is adopting this approach.
For most cases, the best way to handle the failure is to throw.

Unless for rare cases, like this:
aggregatorClazz is a parameter of a URL query, so we need to be defendent in this case.

Comment (manuzhang) [inline core/src/main/scala/io/gearpump/util/HistoryMetricsService.scala]: yes. What I'm concerned is whether we are responsive to users. Does user know what fails and why ?

Comment (clockfly) [inline core/src/main/scala/io/gearpump/util/HistoryMetricsService.scala]: If aggregatorClazz is wrong, what we do now is to log it with error log level. Admin can view the logs to find the errors. 

It is still "responsive", as we will use the SkipAllAggregator to return an empty set.

**Comment by manuzhang** [comment]
+1

**Pull request merged**